src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -293,6 +433,132 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+                    //
+                    //   A -> B -> C
+                    //   A -> D
+                    //
+                    // This case is not hit for project C. The reverse transitive references for C contains B, which is
+                    // the target project of the removed reference. We can see that project C is impacted by this
+                    // removal, and after the removal, project A will no longer be in the reverse transitive references
+                    // of C.
+                    //
+                    // This case is hit for project D. The reverse transitive references for D does not contain B, which
+                    // means project D cannot be "downstream" of the impact of removing the reference A->B. We can see
+                    // that project A will still be in the reverse transitive references of D.
+                    //
+                    // This optimization does not catch all cases. For example, consider the following graph where we
+                    // are removing the project reference A->B:
+                    //
+                    //   A -> B -> D
+                    //   A -> D
+                    //
+                    // For this example, we do not hit this optimization because D contains project B in the set of
+                    // reverse transitive references. Without more complicated checks, we cannot rule out the
+                    // possibility that A may have been removed from the reverse transitive references of D by the
+                    // removal of the A->B reference.
+                    continue;
+                }
+
+                if (references.Contains(projectId))
+                {
+                    builder.Remove(project);
+                }
+            }
+
+            // Invalidate the transitive references from the previously-referenced project
+            builder.Remove(referencedProjectId);
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+        {
+            Contract.ThrowIfFalse(_projectIds.Contains(projectId));
+
+            // Project ID set and direct forward references are trivially updated by removing the key corresponding to
+            // the project getting removed.
+            var projectIds = _projectIds.Remove(projectId);
+            var referencesMap = ComputeNewReferencesMapForRemovedProject(_lazyReverseReferencesMap, _referencesMap, projectId);
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = _lazyReverseReferencesMap is object
+                ? ComputeNewReverseReferencesMapForRemovedProject(_referencesMap, _lazyReverseReferencesMap, projectId)
+                : null;
```

Comment:
same comment as below.  can we push this logic into ComputeNewReverseReferencesMapForRemovedProject?

Reply:
This is the pattern already used elsewhere in this file. Resolving as out of scope for this PR.

Reply:
you're making some fairly big and complex changes here.  I think we should consider these easy to bite-off improvements to make the flow of things clearer IMO.

Reply:
It can be a separate PR dedicated to that specific improvement.

Reply:
Unresolving to either have addressed or done as a mop-up: in review I similarly thought "wait, what's going on here?" and then expanded this and @CyrusNajmabadi saw the same way.

Reply:
Updated locally; will mark resolved on push

Reply:
➡️ Now pushed.